### ISSUE TITLE
Juju charm

### ISSUE THREAD

**Issue by strugee** [pump-io/pump.io#1236]
It would be super neat if we had a Juju charm for deploying pump.io.

**Comment by skellat** [comment]
See: https://jujucharms.com/docs/stable/authors-charm-best-practice

While a nice idea, you have to remember that Juju is an orchestration tool.  Some underlying plumbing has to be available first to be orchestrated.  We're not there yet.

If everything is packaged and in a .deb archive this would work great.  There was a prior effort Debian-side to actually package the bits and pieces of pump.io.  The last update on the relevant wiki is shown from April 2016 here: https://wiki.debian.org/Javascript/Nodejs/Tasks/Pump.io .  I don't think that effort has moved forward.

If anything, we would need to consider Snapcraft (http://snapcraft.io/docs/build-snaps/) or something similar *first* due to the reliance on JS as not all of that can be wrangled into .deb to come into Juju's scope.  

I'll need to think about this further and may expound upon this further later.

**Comment by ssweeny** [comment]
> If anything, we would need to consider Snapcraft

I actually had a go at this a while back. You can find it here: https://github.com/ssweeny/pumpio-snap

It works well enough to get an instance going, but there's room for improvement with e.g. getting TLS certificates.

**Comment by strugee** [comment]
@skellat I don't think we need to be packaged as a .deb first (unless that's a technical requirement?)

We can just pull from npm.

@ssweeny thanks, that's useful. We got contacted by some Canonical developers recently about packaging for Snapcraft; I'm planning to take a look when I have time. If you want to keep working on your packaging I'd be thrilled to accept PRs upstreaming that work here

**Comment by ssweeny** [comment]
@strugee who did you hear from inside Canonical? I work there too (though probably in a different area... snapping pump.io is a personal project) and I'd be happy to chat with them.

**Comment by strugee** [comment]
@ssweeny Daniel Holbach and Alan Pope contacted Evan, who then looped me in.
